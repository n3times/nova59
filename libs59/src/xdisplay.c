#include "s_internal.h"

#include <assert.h>
#include <string.h>

#define CHECK_DISPLAY_EDIT(d) { CHECK(d); assert(check_xdisplay(d)); }


/******************************************************************************
 *
 *  HELPERS.
 *
 ******************************************************************************/

#if !NDEBUG
static bool is_digit(char c) {
  return c >= '0' && c <= '9';
}

static bool check_xdisplay(xdisplay_t * x) {
  char *d = x->display;

  // Should not start with "0d" or "-0d".
  if (*d == '-') d += 1;
  if (*d == '0' && is_digit(*(d + 1))) return false;

  // One dot and up to 8 or 10 digits.
  int digit_count = 0;
  bool has_dot = false;
  while (*d == '.' || is_digit(*d)) {
    if (*d == '.') {
      if (has_dot) return false;
      has_dot = true;
    } else {
      digit_count += 1;
    }
    d += 1;
  }

  // Has no exp.
  bool is_exp = *d != '\0';
  if (!is_exp) {
    if (digit_count > 10) return false;
    if (x->mode == XDISPLAY_MODE_EDIT_EXP) return false;
    return true;
  }

  // Has exp.
  if (digit_count > 8) return false;
  if (strlen(d) != 3) return false;
  if (*d != ' ' && *d != '-') return false;
  if (!is_digit(*(d + 1))) return false;
  if (!is_digit(*(d + 2))) return false;
  return true;
}
#endif

/** Returns the location of the first character after the mantissa. */
static char *get_end_mant(char *start_mant) {
  int len = strlen(start_mant);
  if (len <= 3) return start_mant + len;
  char *c = start_mant + len - 3;
  if (*c == ' ' || *c == '-') return c;
  return start_mant + len;
}

static void insert_at_end_mant(char *start_mant, char c) {
  char *end_mant = get_end_mant(start_mant);

  // Move exponent.
  memmove(end_mant + 1, end_mant, strlen(end_mant) + 1);

  *end_mant = c;
}

/**
 * Updates the display taking into account 'x', 'fix', 'ee' and 'eng'.
 *
 * In case of overflow, sets 'blinking' to true;
 */
static void update_display(xdisplay_t *d) {
  n_format_t format = d->eng ? N_ENG : d->ee ? N_SCI : N_FLOAT;
  n_err_t err;
  n_n2s(d->reg_x, d->fix, format, d->display, &err);
  if (err) d->blinking = true;
}


/******************************************************************************
 *
 *  IMPLEMENTATION.
 *
 ******************************************************************************/

void xdisplay_init(xdisplay_t *d) {
  CHECK(d);

  xdisplay_clear(d);
  d->eng = false;
  d->fix = 9;

  CHECK_DISPLAY_EDIT(d);
}

void xdisplay_clear(xdisplay_t *d) {
  CHECK(d);

  d->display[0] = '0';
  d->display[1] = '\0';
  d->blinking = false;
  d->mode = XDISPLAY_MODE_EDIT_MANT;
  d->reg_x = N_0;
  d->ee = false;

  CHECK_DISPLAY_EDIT(d);
}

void xdisplay_clear_entry(xdisplay_t *d) {
  CHECK(d);

  d->blinking = false;

  if (d->mode != XDISPLAY_MODE_REG) {
    d->display[0] = '0';
    d->display[1] = '\0';
    d->mode = XDISPLAY_MODE_EDIT_MANT;
    d->reg_x = N_0;
  }

  CHECK_DISPLAY_EDIT(d);
}

void xdisplay_digit(xdisplay_t *x, int digit) {
  CHECK_D(digit);
  CHECK_DISPLAY_EDIT(x);
  char *d = x->display;

  if (x->mode == XDISPLAY_MODE_REG) {
    x->display[0] = '0' + digit;
    x->display[1] = '\0';
    x->mode = XDISPLAY_MODE_EDIT_MANT;
    return;
  }

  // Add digit to exponent.
  if (x->mode == XDISPLAY_MODE_EDIT_EXP) {
    int len = strlen(d);
    d[len - 2] = d[len - 1];
    d[len - 1] = '0' + digit;
    return;
  }

  // Add digit to mantissa.

  char *start = *d == '-' ? d + 1 : d;
  char *end = get_end_mant(d);

  // Case "0" or "-0".
  if (*start == '0' && end == start + 1) {
    *start = '0' + digit;
    return;
  }

  // Count mantissa digits.
  int dig_count = 0;
  for (char *p = start; p != end; p++) {
    if (p == end) break;
    if (*p == '.') continue;
    dig_count += 1;
  }

  // See if there is room for digit.
  bool exp = *end != '\0';
  if (dig_count == (exp ? 8 : 10)) {
    if (*start == '0' && start[1] == '.') {
      // Remove 0.
      memmove(start, start + 1, strlen(start + 1) + 1);
    } else {
      // No room for digit.
      return;
    }
  }

  insert_at_end_mant(d, '0' + digit);
}

void xdisplay_dot(xdisplay_t *x) {
  CHECK_DISPLAY_EDIT(x);
  char *d = x->display;

  if (x->mode == XDISPLAY_MODE_REG) {
    x->display[0] = '0';
    x->display[1] = '.';
    x->display[2] = '\0';
    x->mode = XDISPLAY_MODE_EDIT_MANT;
    return;
  }

  x->mode = XDISPLAY_MODE_EDIT_MANT;
  char *dot = strchr(d, '.');
  if (dot) return;
  insert_at_end_mant(d, '.');
}

void xdisplay_chs(xdisplay_t *x) {
  CHECK_DISPLAY_EDIT(x);
  char *d = x->display;

  if (x->mode == XDISPLAY_MODE_EDIT_EXP) {
    char *sign = d + strlen(d) - 3;
    *sign = *sign == '-' ? ' ' : '-';
  } else {
    bool neg = *d == '-';
    if (neg) {
      memmove(d, d + 1, strlen(d + 1) + 1);
    } else {
      memmove(d + 1, d, strlen(d) + 1);
      d[0] = '-';
    }
  }
}

void xdisplay_ee(xdisplay_t *x) {
  CHECK_DISPLAY_EDIT(x);
  char *d = x->display;

  x->ee = true;

  if (x->mode == XDISPLAY_MODE_EDIT_EXP) return;

  // Add exponent if missing.

  char *end_mant = get_end_mant(d);
  if (*end_mant != '\0') {
    x->mode = XDISPLAY_MODE_EDIT_EXP;
    return;
  }

  int dig_count = 0;
  for (char *c = d; *c != '\0'; c++) {
    if (*c >= '0' && *c <= '9') dig_count += 1;
  }

  if (dig_count == 9) {
    int i = 0;
    if (d[0] == '-') i = 1;
    if (d[i] == '0' && d[i + 1] == '.') {
      memmove(d + i, d + i + 1, strlen(d+ i + 1) + 1);
      dig_count = 8;
    }
  }

  if (dig_count <= 8) {
    insert_at_end_mant(d, ' ');
    insert_at_end_mant(d, '0');
    insert_at_end_mant(d, '0');
    x->mode = XDISPLAY_MODE_EDIT_EXP;
  } else {
    x->mode = XDISPLAY_MODE_EDIT_MANT;
  }
}

void xdisplay_iee(xdisplay_t *x) {
  CHECK_DISPLAY_EDIT(x);

  x->ee = false;

  if (x->mode == XDISPLAY_MODE_REG) return;

  x->mode = XDISPLAY_MODE_EDIT_MANT;

  char *d = x->display;
  int len = strlen(d);
  if (len >= 3 && d[len - 3] == ' ' && d[len - 2] == '0' && d[len - 1] == '0') {
    d[len - 3] = '\0';
  }
}

void xdisplay_eng(xdisplay_t *d) {
  d->eng = true;
  d->mode = XDISPLAY_MODE_REG;
  update_display(d);
}

void xdisplay_ieng(xdisplay_t *d) {
  d->eng = false;
  d->mode = XDISPLAY_MODE_REG;
  update_display(d);
}

void xdisplay_fix(xdisplay_t *d, int fix) {
  d->fix = fix;
  d->mode = XDISPLAY_MODE_REG;
  update_display(d);
}

void xdisplay_blink(xdisplay_t *d) {
  d->blinking = true;
}

void xdisplay_update_reg_x(xdisplay_t *d, n_t X) {
  d->mode = XDISPLAY_MODE_REG;
  d->reg_x = X;
  update_display(d);
}

n_t xdisplay_resolve_edit(xdisplay_t *d) {
  assert(d->mode != XDISPLAY_MODE_REG);

  n_err_t err;
  n_t X = n_s2n(d->display, &err);
  d->reg_x = X;
  if (err) d->blinking = true;

  return X;
}
