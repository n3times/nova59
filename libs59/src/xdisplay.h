/**
 * xdisplay.h
 *
 * In charge of displaying and editing register X.
 */

#ifndef XDISPLAY_H
#define XDISPLAY_H

#include "n59.h"


/******************************************************************************
 *
 *  TYPES.
 *
 ******************************************************************************/

/** Indicate whether register X is being displayed or edited. */
typedef enum xdisplay_mode_e {
  XDISPLAY_MODE_REG,             // X is being displayed.
  XDISPLAY_MODE_EDIT_MANT,       // X's mantissa is being edited.
  XDISPLAY_MODE_EDIT_EXP         // X's exponent is being edited.
} xdisplay_mode_t;

/** The display state. */
typedef struct xdisplay_s {
  char display[N_N2S_MAX_SIZE];  // how X, or its editing, appears on display.
  xdisplay_mode_t mode;          // whether X is being edited or displayed.
  bool blinking;                 // whether the display is blinking.

  // Data used to display X in register mode.
  n_t reg_x;                     // register X.
  int fix;                       // fix value.
  bool ee;                       // ee mode.
  bool eng;                      // eng notation.
} xdisplay_t;


/******************************************************************************
 *
 *  INIT & CLEAR.
 *
 ******************************************************************************/

/**
 * Inits display.
 *
 * This function should be called when TI-59 is turned on.
 *
 * Mode: n/a -> XDISPLAY_MODE_EDIT_MANT.
 */
void xdisplay_init(xdisplay_t *x);

/**
 * Resets 'x' except for 'fix' and 'eng'.
 *
 * This function should be called when 'CLR' is pressed.
 *
 * Mode: * -> XDISPLAY_MODE_EDIT_MANT.
 */
void xdisplay_clear(xdisplay_t *x);

/**
 * Sets 'blinking' to false. In addition, in edit mode, resets display to '0'.
 *
 * This function should be called when 'CE' is pressed.
 *
 * Mode: edit     -> XDISPLAY_MODE_EDIT_MANT.
 *       register -> register.
 */
void xdisplay_clear_entry(xdisplay_t *x);

/******************************************************************************
 *
 *  EDIT.
 *
 ******************************************************************************/

/**
 * In edit mode, adds 'd' to display if possible.
 * In register mode, sets display to 'd'.
 *
 * This function should be called when a digit is pressed.
 *
 * Mode: * -> edit.
 */
void xdisplay_digit(xdisplay_t *x, int d);

/**
 * In edit mode, adds '.' to display, if not already present.
 * In register mode, sets display to '0.'.
 *
 * This function should be called when '.' is pressed.
 *
 * Mode: * -> XDISPLAY_MODE_EDIT_MANT.
 */
void xdisplay_dot(xdisplay_t *x);

/**
 * Changes the sign of the mantissa or exponent.
 *
 * This function should be called when '+/-' is pressed. In register mode a
 * separate function ('s_math_chs') should also be called to change the sign of
 * X.
 *
 * Typically when a function modifies X, 'xdisplay_update_reg_x' should be
 * called but not in this case, as we want to show '-0' and not '0', for
 * example.
 *
 * Mode: unchanged.
 */
void xdisplay_chs(xdisplay_t *x);

/******************************************************************************
 *
 *  MODES.
 *
 ******************************************************************************/

/**
 * Sets fix and updates display.
 *
 * Mode * -> register.
 */
void xdisplay_fix(xdisplay_t *x, int fix);

/**
 * Sets scientific mode. Additionally, sets display to edit mode and possibly
 * adds exponent.
 *
 * If there is no exponent and there is enough space, adds exponent '00'. If
 * exponent is added or is already present, sets mode to XDISPLAY_MODE_EDIT_EXP.
 * Otherwise, sets mode to XDISPLAY_MODE_EDIT_MANT.
 *
 * This should be called when 'EE' is pressed.
 *
 * Mode: * -> edit.
 */
void xdisplay_ee(xdisplay_t *x);

/**
 * Unsets scientific mode.
 *
 * In edit mode, sets 'mode' to XDISPLAY_MODE_EDIT_MANT.
 *
 * This should be called when 'INV EE' is pressed.
 *
 * Mode: edit     -> XDISPLAY_MODE_EDIT_MANT.
 *       register -> register.
 */
void xdisplay_iee(xdisplay_t *x);

/**
 * Sets engineering notation, possibly updating display.
 *
 * This should be called when 'Eng' is pressed.
 *
 * Mode: * -> register.
 */
void xdisplay_eng(xdisplay_t *x);

/**
 * Unsets engineering notation, possibly updating display.
 *
 * This should be called when 'INV Eng' is pressed.
 *
 * Mode: * -> register.
 */
void xdisplay_ieng(xdisplay_t *x);

/******************************************************************************
 *
 * SYNCHRONIZATION WITH REGISTER X.
 *
 ******************************************************************************/

/**
 * Sets 'blinking' to true.
 *
 * This function should be called in case of error.
 *
 * Examples: 'STO STO', '0 lnx'.
 *
 * Mode: unchanged.
 */
void xdisplay_blink(xdisplay_t *x);

/**
 * Updates xdisplay state based on reg_x.
 *
 * This function should be called when register X has been modified.
 *
 * Examples: 'lnx', 'RCL 59'.
 * Non examples:
 * - editing keys: '.', 'EE' or '0'.
 * - keys that keep display in edit mode: 'Deg', 'Lbl' or 'RST'.
 * - keys/operations that don't change X: 'STO 59', 'CP'
 *
 * Note: this function should not be called when calling 'xdisplay_chs'.
 *
 * Mode: * -> register.
 */
void xdisplay_update_reg_x(xdisplay_t *x, n_t X);

/**
 * In edit mode, updates reg_x based on the number on the display and returns
 * the new value of reg_x. In addition puts 'x' in register mode.
 *
 * This function should be called when the display is in edit mode and the user
 * is done editing it.
 *
 * The user is done editing the display when one of several keys is pressed such
 * as 'lnx', 'STO' and 'CP'. On the other hand, keys such as 'Lbl', 'Deg' and
 * '+/-', do not end editing mode.
 *
 * Note that this function does not change the display itself. To do so, the
 * caller of this function needs to call 'xdisplay_update_reg_x'. For example
 * for 'lnx':
 *   X = xdisplay_resolve_edit(&x);
 *   s_math_ln(&X, &err);
 *   xdisplay_update_reg_x(&x, X);
 *   if (err) xdisplay_blink(&x);
 *
 * Mode: edit     -> edit.
 *       register -> abort.
 */
n_t xdisplay_resolve_edit(xdisplay_t *x);

#endif  // XDISPLAY_H
