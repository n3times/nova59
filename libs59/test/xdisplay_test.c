#include "xdisplay.h"

#include <assert.h>
#include <stdio.h>

static void input(xdisplay_t *d, char * input) {
  for (char *c = input; *c != '\0'; c++) {
    if (*c == 'k') {
      xdisplay_edit_init(d);
    } else if (*c == '-') {
      xdisplay_edit_chs(d);
    } else if (*c == '.') {
      xdisplay_edit_dot(d);
    } else if (*c == 'e') {
      xdisplay_mode_ee(d);
    } else if (*c == 'i') {
      xdisplay_mode_iee(d);
    } else if (*c == 'E') {
      xdisplay_mode_eng(d);
    } else if (*c == 'I') {
      xdisplay_mode_ieng(d);
    } else if (*c >= '0' && *c <= '9') {
      xdisplay_edit_digit(d, *c - '0');
    }
  }
}

static void test(char *chars) {
  xdisplay_t d;
  xdisplay_edit_init(&d);
  input(&d, chars);
  printf("%s\n", d.display);
}

int main() {
  // "0".
  test("0");
  test("00");
  test("000");
  test("--");
  test("0ei");
  test("-0ei-");
  test("e-ie-i");
  test("1.23e45k");
  test("k");

  // "-0".
  printf("\n");
  test("-");
  test("---");
  test("-0ei");
  test("0-");

  // "0.".
  printf("\n");
  test(".");
  test("..");
  test("0.");
  test("0...--");

  // "-1.2".
  printf("\n");
  test("-1.2");
  test("1.2-");
  test("1-.2");
  test("1.-2");
  test("-1..2");
  test("-1.2.");


  // ".9999999999".
  printf("\n");
  test(".9999999999");
  test("0.9999999999");
  test("000.9999999999");
  test("0.999999999999");
  test(".999999999e9");
  test(".9999999999e");

  // "9999999999".
  printf("\n");
  test("9999999999");
  test("99999999999999");
  test("9999999999e12");
  test("9999999999e12i45");

  // "1 12".
  printf("\n");
  test("1e12");
  test("1ee12");
  test("e12i1");
  test("e11i1e12");
  test("e11i1e1341312");

  // "1234.5678-59".
  printf("\n");
  test("1234.5678e-59");
  test("e-59i1234.5678");
  test("e-59i1234e.5678");

  // "0 00".
  printf("\n");
  test("e");
  test("e--");
  test("-e-i-e-");
  test("0e00");
  test("e0");

  // "-9999999999-99"
  printf("\n");
  test("-99999999.e-99");
  test("99999999e99-.-");
  test("-99999999.e-99i9999");

  // "-.9999999999-99"
  printf("\n");
  test("-.99999999e-99");
  test(".99999999e99-.-");

  // Edit number.
  printf("\n");
  xdisplay_t d;
  xdisplay_edit_init(&d);
  xdisplay_update_x(&d, n_make(1e45));
  printf("%s\n", d.display);
  xdisplay_mode_ee(&d);
  xdisplay_edit_digit(&d, 9);
  xdisplay_edit_dot(&d);
  xdisplay_edit_digit(&d, 7);
  printf("%s\n", d.display);

  // Replace last visible digit of number.
  printf("\n");
  xdisplay_edit_init(&d);
  xdisplay_mode_fix(&d, 4);
  xdisplay_update_x(&d, n_make(12345.54321));
  printf("%s\n", d.display);
  xdisplay_mode_ee(&d);
  printf("%s\n", d.display);
  xdisplay_edit_digit(&d, 5);
  printf("%s\n", d.display);

  // Only keep visible digits of pi.
  printf("\n");
  for (int i = 0; i <= 9; i++) {
    xdisplay_edit_init(&d);
    xdisplay_mode_fix(&d, i);
    xdisplay_update_x(&d, N_PI);
    xdisplay_mode_ee(&d);
    xdisplay_mode_iee(&d);
    n_t X = xdisplay_resolve_edit(&d);
    char str[N_PRINT_MAX_SIZE];
    printf("%s : %s\n", n_print(X, str), d.display);
  }

  // '0 = +/-'.
  printf("\n");
  xdisplay_edit_init(&d);
  xdisplay_update_x(&d, N_0);
  xdisplay_edit_chs(&d);
  printf("%s\n", d.display);

  // 'Fix 0 9.9 EE 99 lnx'.
  printf("\n");
  xdisplay_edit_init(&d);
  xdisplay_mode_fix(&d, 0);
  input(&d, "9.9e99");
  n_t X = xdisplay_resolve_edit(&d);
  n_err_t err;
  X = n_ln(X, &err);
  xdisplay_update_x(&d, X);
  printf("%s\n", d.display);
  assert(!d.blink);

  return 0;
}
