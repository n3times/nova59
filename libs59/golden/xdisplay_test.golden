0
0
0
0
0
0
0
0
0

-0
-0
-0
-0

0.
0.
0.
0.

-1.2
-1.2
-1.2
-1.2
-1.2
-1.2

.9999999999
.9999999999
.9999999999
.9999999999
.9999999999
.9999999999

9999999999
9999999999
9999999999
9999999999

1 12
1 12
1 12
1 12
1 12

1234.5678-59
1234.5678-59
1234.5678-59

0 00
0 00
0 00
0 00
0 00

-99999999.-99
-99999999.-99
-99999999.-99

-.99999999-99
-.99999999-99

1. 45
1.7 59

12345.5432
12345.5432
12345.54325

 3000000000000 00 : 3.
 3100000000000 00 : 3.1
 3140000000000 00 : 3.14
 3142000000000 00 : 3.142
 3141600000000 00 : 3.1416
 3141590000000 00 : 3.14159
 3141593000000 00 : 3.141593
 3141592700000 00 : 3.1415927
 3141592650000 00 : 3.14159265
 3141592654000 00 : 3.141592654

-0.

2. 02
